Enable gcafsv1 aerosol GRIB2 processing in UPP

zhanglikate opened the pull request:
# Description

This PR updates the UPP system to support aerosol GRIB2 output for the gcafsv1 model within UFSATM.

Specifically:

Updates UPP to the develop branch to include the text file required for the GCAFS model in UFSATM.

Adds the post_itag_gcafs file for the GCAFS model.

Implements and links UPP-related components for the gcafsv1 model to process aerosol GRIB2 outputs.

These changes enhance UFSATM’s capability to handle aerosol fields generated by GCAFS for post-processing and analysis.


# Type of change
- [x] New feature (adds functionality)

# Change characteristics
<!-- Choose YES or NO from each of the following and delete the other -->
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES
- Does this change require an update to any of the following submodules? YES (If YES, .)
  - [X] UFS-weather-model <!-- ufs-community/ufs-weather-model#https://github.com/ufs-community/ufs-weather-model/pull/2957 -->


# How has this been tested?
<!-- Please list any test you conducted, including the machine.

Example:
- Clone and build on and Cycled test on URSA
-->

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [x] I have made corresponding changes to the system documentation if necessary

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -592,10 +592,16 @@ FV3_predet(){
+        cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt"      "${DATA}/postxconfig-NT.txt"
+        cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt"      "${DATA}/postxconfig-NT_FH00.txt"
+        cpreq "${PARMgfs}/ufs/post_itag_gcafs"                      "${DATA}/itag"  # TODO: Need a GCAFS version when available in the UFS-weather-model
+        cpreq "${PARMgfs}/post/optics_luts_*.dat"                   "${DATA}/."
```
`*` will not be expanded as expected when placed inside `"*"`.  This line will fail.

zhanglikate replied:
Thanks for the suggestion, though it is working at URSA, I updated as:
        cpreq "${PARMgfs}/post/optics_luts_DUST_nasa.dat"           "${DATA}/."
        cpreq "${PARMgfs}/post/optics_luts_NITR_nasa.dat"           "${DATA}/."
        cpreq "${PARMgfs}/post/optics_luts_SALT_nasa.dat"           "${DATA}/."
        cpreq "${PARMgfs}/post/optics_luts_SOOT_nasa.dat"           "${DATA}/."
        cpreq "${PARMgfs}/post/optics_luts_SUSO_nasa.dat"           "${DATA}/."
        cpreq "${PARMgfs}/post/optics_luts_WASO_nasa.dat"           "${DATA}/."

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -150,11 +150,19 @@ ${LINK_OR_COPY} "${HOMEgfs}/sorc/ufs_model.fd/tests/parm/noahmptable.tbl" .
+    optics_luts_SALT.dat optics_luts_SALT_nasa.dat optics_luts_SOOT.dat optics_luts_SOOT_nasa.dat \
+    optics_luts_SUSO.dat optics_luts_SUSO_nasa.dat optics_luts_WASO.dat optics_luts_WASO_nasa.dat
+do
+  ${LINK_OR_COPY} "${HOMEgfs}/sorc/upp.fd/fix/chem/${file}" .
```
These files should be in the `fix` directory instead of `upp` source code.

WenMeng-NOAA replied:
upp.fd is a clone of the UPP which includes sorc/, fix/, parm/.

aerorahul replied:
Ok.

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -113,6 +113,12 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
+      if [[ ${last} -gt ${ncount} ]]; then
+        last=${ncount}
+      fi
+
```
```suggestion
```
Instead, on line 123 do the following:
```
 if [[ ${iproc} -eq ${nproc} || ${last} -gt ${ncount} ]]; then
```

zhanglikate replied:
Thanks for the suggestion. I have updated it as you suggested.

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -148,13 +158,23 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
   mv mpmd.out "mpmd_${nset}.out"
 
+  # ---- FIX 3a: default nproc_used when every rank produced work ----
+    : "${nproc_used:=${nproc}}"
```
This line is incorrect, it starts with `:`
```suggestion
```

zhanglikate replied:
Deleted .

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -148,13 +158,23 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
+#      rm  -f "pgb2${grp}file_${fhr3}_${iproc}_${grid}"
+       file="pgb2${grp}file_${fhr3}_${iproc}_${grid}"
+       # only concatenate if file exists and is non-empty
+       if [[ -s "$file" ]]; then
```
```suggestion
       if [[ -s "${file}" ]]; then
```

zhanglikate replied:
Updated.

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -148,13 +158,23 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
-      rm  -f "pgb2${grp}file_${fhr3}_${iproc}_${grid}"
+   # ---- FIX 3b: only append files that exist and are non-empty ----
+#      cat "pgb2${grp}file_${fhr3}_${iproc}_${grid}" >> "pgb2${grp}file_${fhr3}_${grid}"
+#      rm  -f "pgb2${grp}file_${fhr3}_${iproc}_${grid}"
```
```suggestion
    # only append files that exist and are non-empty
```
Check indent here after commit

zhanglikate replied:
Updated

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -148,13 +158,23 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
+       file="pgb2${grp}file_${fhr3}_${iproc}_${grid}"
+       # only concatenate if file exists and is non-empty
+       if [[ -s "$file" ]]; then
+       cat "$file" >> "pgb2${grp}file_${fhr3}_${grid}"
```
```suggestion
       cat "${file}" >> "pgb2${grp}file_${fhr3}_${grid}"
       rm -f "${file}"
```

zhanglikate replied:
Updated

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -148,13 +158,23 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
+       if [[ -s "$file" ]]; then
+       cat "$file" >> "pgb2${grp}file_${fhr3}_${grid}"
+       fi
+       rm -f -- "$file"
```
```suggestion
```

zhanglikate replied:
Updated

aerorahul commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -131,7 +137,11 @@ for (( nset=1 ; nset <= downset ; nset++ )); do
+  # FIX 2a: record how many processors actually did work
+      nproc_used=${iproc}
+ # FIX 2b: pad the rest with dummy lines (<= instead of <)
+      #for (( pproc = iproc+1 ; pproc < nproc ; pproc++ )); do
```
```suggestion
```
remove commented code (agree on the bug here)

zhanglikate replied:
Remove the commented code, but " nproc_used=${iproc}" need to be kept

aerorahul commented on dev/ci/cases/yamls/gcafs_cycled_noDA_defaults_dev.yaml:
```diff
@@ -1,9 +1,9 @@
   FHMAX: 24
-  FHOUT_AERO: 1
\ No newline at end of file
+  FHOUT_AERO: 1
```
```suggestion
  FHOUT_AERO: 1

```

aerorahul reviewed (changes requested):
Thanks for the PR.
Some comments

WenMeng-NOAA commented on parm/product/gcafs.fFFF.paramlist.b.txt:
```diff
@@ -0,0 +1,349 @@
+ABSV:125 mb
```
@zhanglikate Currently, the UPP control files for gcafs only include aerosols fields (See [here](https://github.com/NOAA-EMC/UPP/blob/develop/parm/gcafs/postcntrl_gcafs.xml)). For the gcafs case, generating dataset a should be sufficient.

bbakernoaa replied:
@zhanglikate I suggest that we revisit this as for production we will need to keep the current separation of met fields and aerosol products for dissemination of products for NOMADS.  Let's make a separate PR to generate met products for out internal use later.

zhanglikate replied:
@WenMeng-NOAA Currently,  for the file "gcafs.fFFF.paramlist.b.txt" (Met. fields output ), I just copied it from gfs model, EMC (@bbakernoaa @lipan-NOAA ) may make further changes in the future when they finalize the Met. output for GCAFS model.

aerorahul commented:
@zhanglikate
Can you please merge `develop` in this branch, resolve conflicts and push?  This will help with review and testing.
Thanks.

aerorahul commented:
Adding @JessicaMeixner-NOAA as a reviewer since this PR contains a `ufs_model.fd` submodule update.

JessicaMeixner-NOAA reviewed (commented):
I was asked to review because of the update of hash in ufs-weather-model.  I do not think that this PR will work without further updates that are associated with various ufs-weather-model updates.   Have the CI tests been run with this?  My guess is that they will fail.  

I plan to work on a PR with updates from UFS hashes.  I was going to wait until the PR that is being tested today in ufs was merged because that's what we need for GFSv17.  I can put together a branch with those updates this afternoon without that, to help expedite this PR if waiting until Monday is not okay.

zhanglikate commented:
> Adding @JessicaMeixner-NOAA as a reviewer since this PR contains a `ufs_model.fd` submodule update.

The related update about this PR in WM has been updated and merged: https://github.com/ufs-community/ufs-weather-model/pull/2957#issuecomment-3472930206
@JessicaMeixner-NOAA @aerorahul

JessicaMeixner-NOAA commented:
The testing section of your PR is empty, how have you tested your PR here? Is it working without additional modifications to the workflow?

zhanglikate commented:
> The testing section of your PR is empty, how have you tested your PR here? Is it working without additional modifications to the workflow?

For the WM part (ufs_model.fd), it has passed all the RT.

zhanglikate commented:
> @zhanglikate Can you please merge `develop` in this branch, resolve conflicts and push? This will help with review and testing. Thanks.

@aerorahul I have merged it with develop branch, thanks.

JessicaMeixner-NOAA commented:
Updates required to make the model run can be found here: https://github.com/NOAA-EMC/global-workflow/pull/4210 I plan to wait until the current PR in UFS that's being tested is merged before opening the PR.  That should be Monday, but its here for reference in the meantime.

JessicaMeixner-NOAA commented:
> > The testing section of your PR is empty, how have you tested your PR here? Is it working without additional modifications to the workflow?
> 
> For the WM part (ufs_model.fd), it has passed all the RT.

@zhanglikate  - Do you mean the ufs-weather-model RT or the global-worklfow CI tests?   Passing the ufs-weather-model RT's is not sufficient to ensure the global-workflow will run the model.  That being said, https://github.com/NOAA-EMC/global-workflow/pull/4210 has been merged.  If you merge that in your PR here this should be fine.

zhanglikate commented:
> > > The testing section of your PR is empty, how have you tested your PR here? Is it working without additional modifications to the workflow?
> > 
> > 
> > For the WM part (ufs_model.fd), it has passed all the RT.
> 
> @zhanglikate - Do you mean the ufs-weather-model RT or the global-worklfow CI tests? Passing the ufs-weather-model RT's is not sufficient to ensure the global-workflow will run the model. That being said, #4210 has been merged. If you merge that in your PR here this should be fine.

@JessicaMeixner-NOAA please check https://github.com/ufs-community/ufs-weather-model/pull/2957#pullrequestreview-3390783538 and https://github.com/ufs-community/ufs-weather-model/pull/2942 , I was following the guidance to finish the RT run and push to the PR in WM.

JessicaMeixner-NOAA reviewed (commented):
@zhanglikate  - Can you update the hash that ufs-weather-model is pointing to here? I think you should be using what is currently in the develop of global-workflow and not another hash.

zhanglikate commented:
> @zhanglikate - Can you update the hash that ufs-weather-model is pointing to here? I think you should be using what is currently in the develop of global-workflow and not another hash.

@JessicaMeixner-NOAA I was originally using the global workflow hash for ufs-weather-model, when I submit the PR to ufs-weather-model, the code manager asked me to use the most recent hash for ufs-weather-model and run the RT https://github.com/NOAA-EMC/global-workflow/pull/4210. Actually, I only added one file into the ufs-weather-model which will not impact on other model's configurations, however, it links to all the updates in these global workflow PR. Please guide me to what is the best way to proceed this global workflow PR. Thanks very much.

JessicaMeixner-NOAA commented:
Right now, you are not using the same ufs-waether-model hash as what was used in #4210.   Seems like things got unintentionally out of sync.   You just need to update your hash of ufs-weather-model to what is being used in global-workflow develop right now and you'll be good to go.   You can confirm you've done that when you look at the files changed in this PR and you do not see ufs_model.fd as being changed.

zhanglikate commented:
> Right now, you are not using the same ufs-waether-model hash as what was used in #4210. Seems like things got unintentionally out of sync. You just need to update your hash of ufs-weather-model to what is being used in global-workflow develop right now and you'll be good to go. You can confirm you've done that when you look at the files changed in this PR and you do not see ufs_model.fd as being changed.

@JessicaMeixner-NOAA I revert the ufs-weather-model to what is being used in global-workflow develop, can you please help to check? Thanks.

JessicaMeixner-NOAA commented:
@zhanglikate  - The ufs-weather-model hash is now as expected.

zhanglikate commented:
@aerorahul  HI Rahul, I have pushed a version to solve the conflicts ([159c61f](https://github.com/NOAA-EMC/global-workflow/pull/4147/commits/159c61fa7c43935facae85335a5f97e6e6d9a4fc)), that what it always shows "This branch has conflicts that must be resolved"? Thanks.

aerorahul commented:
@zhanglikate, if you click on resolve conflicts, make updates and push them, the conflicts will disappear.

zhanglikate commented:
> @zhanglikate, if you click on resolve conflicts, make updates and push them, the conflicts will disappear.

But none of the options is correct. What is the best for me to do that? Do you have any suggestions?

aerorahul commented:
You might consider checking out the `develop` version of the script, and apply the changes from your branch, by cherry picking only those that are intentional.

zhanglikate commented:
@aerorahul  Hi, I have merged this script with the develop branch and push it. Do you think it is good now, or I need to merge everything with the develop branch again? Thanks.

CoryMartin-NOAA commented:
@zhanglikate I see you've force pushed, but there are still conflicts showing here on the GitHub PR.

zhanglikate commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -115,8 +116,14 @@
     # if at final record and have not reached the final processor then write echo's to
     # cmdfile for remaining processors
     if [[ ${last} -eq ${ncount} ]]; then
+<<<<<<< HEAD
```
Solved.

zhanglikate commented on scripts/exglobal_atmos_products.sh:
```diff
@@ -115,8 +116,14 @@
     # if at final record and have not reached the final processor then write echo's to
     # cmdfile for remaining processors
     if [[ ${last} -eq ${ncount} ]]; then
+<<<<<<< HEAD
```
Solved.

zhanglikate commented:
> @zhanglikate I see you've force pushed, but there are still conflicts showing here on the GitHub PR.

@CoryMartin-NOAA Hi Cory, do you see which file has conflict? Do you think I need to merge the whole package with develop again to solve it?

CoryMartin-NOAA commented:
<img width="931" height="671" alt="image" src="https://github.com/user-attachments/assets/f2439407-41c2-4151-8357-8354a4bbfd2a" />

aerorahul commented:
@zhanglikate 
`scripts/` is now a soft-link to `dev/scripts/` 
I think the conflict is that you have created a file `scripts/exglobal_atmos_products.sh` which conflicts with the `dev/scripts/exglobal_atmos_products.sh`.

zhanglikate commented:
> @zhanglikate `scripts/` is now a soft-link to `dev/scripts/` I think the conflict is that you have created a file `scripts/exglobal_atmos_products.sh` which conflicts with the `dev/scripts/exglobal_atmos_products.sh`.

Oh, I see. But my current version, does not have dev/scripts, so I may need to merge the whole package with the develop branch again? Thanks.

aerorahul commented:
I looked at your branch and it is not rebased on `NOAA-EMC/global-workflow::develop`.
There are a lot of differences that do not include the changes from a recent PR #4259 
I also notice that your fork is a fork from `jkhenderson/global-workflow` which is significantly out of date.

zhanglikate commented:
> I looked at your branch and it is not rebased on `NOAA-EMC/global-workflow::develop`. There are a lot of differences that do not include the changes from a recent PR #4259 I also notice that your fork is a fork from `jkhenderson/global-workflow` which is significantly out of date.

@aerorahul The fork i can not change because my global workflow was forked several years ago, but it should impact the PR. My branch is from recent global workflow (August version) to run GCAFS model and I have merged with the develop branch at the end of Oct 31. I agree, it is out of date compare to today's develop branch, that is why I am asking should I need to merge with the develop branch again? Thanks.

Kate

aerorahul commented:
> > I looked at your branch and it is not rebased on `NOAA-EMC/global-workflow::develop`. There are a lot of differences that do not include the changes from a recent PR #4259 I also notice that your fork is a fork from `jkhenderson/global-workflow` which is significantly out of date.
> 
> @aerorahul The fork i can not change because my global workflow was forked several years ago, but it should impact the PR. My branch is from recent global workflow (August version) to run GCAFS model and I have merged with the develop branch at the end of Oct 31. I agree, it is out of date compare to today's develop branch, that is why I am asking should I need to merge with the develop branch again? Thanks.
> 
> Kate

Yes. You should merge upstream `develop` into your feature branch everytime the upstream `develop` is updated. Please do this again now and up until the PR is merged.

bbakernoaa commented:
@zhanglikate @aerorahul point is that there have been massive amounts of changes since then.  It may be easier to create a new branch directly off the develop branch of global workflow 

Here is a series of commands for you to execute to accomplish this 

```bash
git remote add emc https://github.com/noaa-emc/global-workflow
git fetch emc 
git checkout emc/develop
```

then to add that branch back to your fork 
```bash
git branch feature/kates_stuff
git checkout feature/kates_stuff

# add your changes 
git add (your files) 
git commit -m 'message here'
git push 
```

zhanglikate commented:
> @zhanglikate @aerorahul point is that there have been massive amounts of changes since then. It may be easier to create a new branch directly off the develop branch of global workflow
> 
> Here is a series of commands for you to execute to accomplish this
> 
> ```shell
> git remote add emc https://github.com/noaa-emc/global-workflow
> git fetch emc 
> git checkout emc/develop
> ```
> 
> then to add that branch back to your fork
> 
> ```shell
> git branch feature/kates_stuff
> git checkout feature/kates_stuff
> 
> # add your changes 
> git add (your files) 
> git commit -m 'message here'
> git push 
> ```

@bbakernoaa @aerorahul  Thanks very much for your guidance. I will following @bbakernoaa 's suggestion to update my PR.

CoryMartin-NOAA commented:
@zhanglikate I now see a lot of files that are conflicts...

zhanglikate commented:
I am still working on it, have not yet finished. Looks like the best way is
adding my changes into the develop branch as the code structure has been
changed a lot in the current develop branch.

On Wed, Dec 10, 2025 at 2:08 PM Cory Martin ***@***.***>
wrote:

> *CoryMartin-NOAA* left a comment (NOAA-EMC/global-workflow#4147)
> <https://github.com/NOAA-EMC/global-workflow/pull/4147#issuecomment-3638965875>
>
> @zhanglikate <https://github.com/zhanglikate> I now see a lot of files
> that are conflicts...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/4147#issuecomment-3638965875>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APJPDRBSCTHKL54C3B2FAK34BCDURAVCNFSM6AAAAACI3ZZ5OKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTMMZYHE3DKOBXGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zhanglikate committed:
Update the UPP for GCAFS model
1)Include the text file required for the GCAFS model in UFSATM.
2) Add the post_itag_gcafs file for the GCAFS model.
3) Implement and link UPP-related files for the gcafsv1 model
to process aerosol GRIB2 outputs.
4) Fix a typo in gcafs yaml file
5) Update config.atmos_products for GCAFS model

zhanglikate commented:
@aerorahul I have pushed a new version of the PR. Because the current develop branch has undergone major structural changes, I rebuilt my PR-upp branch on top of the latest develop and reapplied only the intended updates. This approach avoids the large number of conflicts that occurred previously.
Please review and merge these updates into the develop branch. Let me know if you have any questions or need additional adjustments.

Thanks for all of your helps and suggestions: @aerorahul @JessicaMeixner-NOAA @CoryMartin-NOAA @bbakernoaa @ypwang19

bbakernoaa committed:
Apply suggestions from code review

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

bbakernoaa commented:
@zhanglikate this looks much better. No conflicts and only minimal changes. Thanks for going taking the extra effort to get this through.

CoryMartin-NOAA commented:
Looks like we still (or now again) have one conflict

bbakernoaa committed:
Merge branch 'develop' into PR-upp

bbakernoaa commented:
@DavidHuber-NOAA @TerrenceMcGuinness-NOAA the `shfmt` CI job failed.  This shouldn't be related to this PR

aerorahul commented on dev/scripts/exglobal_atmos_products.sh:
```diff
@@ -114,7 +114,8 @@ for ((nset = 1; nset <= downset; nset++)); do
         # cmdfile for remaining processors
         if [[ "${last}" -eq "${ncount}" ]]; then
-            for ((pproc = iproc + 1; pproc < nproc; pproc++)); do
+            nproc_used=${iproc}
```
Where is `nproc_used` used?

zhanglikate replied:
Deleted.

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -556,10 +556,21 @@ FV3_predet() {
-                ;;
+            gfs)
+              cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-two.txt"     "${DATA}/postxconfig-NT.txt"
+              cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-f00-two.txt" "${DATA}/postxconfig-NT_FH00.txt"
```
This (and below) should be indented 4 spaces to match the rest of the code in this file.

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -556,10 +556,21 @@ FV3_predet() {
+              cpreq "${PARMgfs}/post/optics_luts_SALT_nasa.dat"           "${DATA}/."
+              cpreq "${PARMgfs}/post/optics_luts_SOOT_nasa.dat"           "${DATA}/."
+              cpreq "${PARMgfs}/post/optics_luts_SUSO_nasa.dat"           "${DATA}/."
+              cpreq "${PARMgfs}/post/optics_luts_WASO_nasa.dat"           "${DATA}/."
```
`shfmt` will fail here for the extra space between the source and target.

bbakernoaa replied:
```suggestion
                cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-two.txt" "${DATA}/postxconfig-NT.txt"
                cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-f00-two.txt" "${DATA}/postxconfig-NT_FH00.txt"
              ;;
            gcafs)
                cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt" "${DATA}/postxconfig-NT.txt"
                cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt" "${DATA}/postxconfig-NT_FH00.txt"
                cpreq "${PARMgfs}/ufs/post_itag_gcafs" "${DATA}/itag"  # TODO: Need a GCAFS version when available in the UFS-weather-model
                cpreq "${PARMgfs}/post/optics_luts_DUST_nasa.dat" "${DATA}/."
                cpreq "${PARMgfs}/post/optics_luts_NITR_nasa.dat" "${DATA}/."
                cpreq "${PARMgfs}/post/optics_luts_SALT_nasa.dat" "${DATA}/."
                cpreq "${PARMgfs}/post/optics_luts_SOOT_nasa.dat" "${DATA}/."
                cpreq "${PARMgfs}/post/optics_luts_SUSO_nasa.dat" "${DATA}/."
                cpreq "${PARMgfs}/post/optics_luts_WASO_nasa.dat" "${DATA}/."
```

bbakernoaa committed:
Apply suggestion from @bbakernoaa

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -181,6 +189,7 @@ declare -a ufs_templates=("model_configure.IN" "input_global_nest.nml.IN"
   "ufs.configure.leapfrog_atm_wav.IN"
   "ww3_shel.nml.IN"
   "post_itag_gfs"
+  "post_itag_gcafs"
```
I see this `post_itag_gcafs` being linked, but the `forecast_predet.sh` is still using `post_itag_gfs` (see line 556 in forecast_predet.sh)

zhanglikate replied:
This PR does not touch the gfs model file of post_itag_gfs in both link_workflow.sh and forecast_predet.sh. We only link post_itag_gcafs once for gcafs model

aerorahul replied:
Resolved as `post_itag_gcafs` is being used in `forecast_predet.sh`.  Thanks @bbakernoaa

aerorahul commented:
> @DavidHuber-NOAA @TerrenceMcGuinness-NOAA the `shfmt` CI job failed. This shouldn't be related to this PR

The linter is back up and running.

zhanglikate commented:
> Looks like we still (or now again) have one conflict

Yes. There is new commit to the workflow after I pushed my PR.

zhanglikate committed:
Remove nproc_used=${iproc} in exglobal_atmos_products.sh

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -556,10 +556,21 @@ FV3_predet() {
                 cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-two.txt" "${DATA}/postxconfig-NT.txt"
                 cpreq "${PARMgfs}/post/gfs/postxconfig-NT-gfs-f00-two.txt" "${DATA}/postxconfig-NT_FH00.txt"
-                ;;
+              ;;
```
```suggestion
                ;;
```

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -556,10 +556,21 @@ FV3_predet() {
+                cpreq "${PARMgfs}/post/optics_luts_SOOT_nasa.dat" "${DATA}/."
+                cpreq "${PARMgfs}/post/optics_luts_SUSO_nasa.dat" "${DATA}/."
+                cpreq "${PARMgfs}/post/optics_luts_WASO_nasa.dat" "${DATA}/."
+              ;;
```
```suggestion
                ;;
```

aerorahul commented on ush/forecast_predet.sh:
```diff
@@ -556,10 +556,21 @@ FV3_predet() {
+            gcafs)
+                cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt" "${DATA}/postxconfig-NT.txt"
+                cpreq "${PARMgfs}/post/gcafs/postxconfig-NT-gcafs.txt" "${DATA}/postxconfig-NT_FH00.txt"
+                cpreq "${PARMgfs}/ufs/post_itag_gcafs" "${DATA}/itag"  # TODO: Need a GCAFS version when available in the UFS-weather-model
```
```suggestion
                cpreq "${PARMgfs}/ufs/post_itag_gcafs" "${DATA}/itag"
```
TODO is not necessary as the file is being linked from the ufs-weather-model in `link_workflow.sh`

zhanglikate replied:
@aerorahul  We used the same way as the GFS model for "post_itag_gfs" , which is also has been linked in link_workflow.sh. I think we need both of them, the link_workflow.sh is linking the post_itag_gcafs from ufs-weather-model to ${PARMgfs}/ufs/ , while here the "cpreq "${PARMgfs}/ufs/post_itag_gcafs" "${DATA}/" is copy the post_itag_gcafs to RUNDIR as itag. So we can not remove the #TODO

aerorahul committed:
Update ush/forecast_predet.sh

aerorahul committed:
Update ush/forecast_predet.sh

aerorahul committed:
Update ush/forecast_predet.sh

aerorahul reviewed (dismissed):
looks good now.
@bbakernoaa @CoryMartin-NOAA 
If you approve, I am going to trigger CI tests on this PR right away.

TerrenceMcGuinness-NOAA commented:
The Ursa build for this PR is failing:
```
12/11/25 20:29:32 UTC :: build.xml :: database or disk is full
.FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
Rocoto reported that the build failed for gsi_enkf
Rocoto reported that the build failed for gfs_utils
.FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
Rocoto reported that the build failed for gsi_enkf
Rocoto reported that the build failed for gfs_utils
.FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
Rocoto reported that the build failed for gsi_enkf
Rocoto reported that the build failed for gfs_utils
.FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
Rocoto reported that the build failed for gsi_enkf
Rocoto reported that the build failed for gfs_utils
```
GitLab pipeline[ build link](https://vlab.noaa.gov/gitlab-community/NWS/Operations/NCEP/EMC/global-workflow/-/jobs/39970)

zhanglikate commented:
> The Ursa build for this PR is failing:
> 
> ```
> 12/11/25 20:29:32 UTC :: build.xml :: database or disk is full
> .FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
> Rocoto reported that the build failed for gsi_enkf
> Rocoto reported that the build failed for gfs_utils
> .FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
> Rocoto reported that the build failed for gsi_enkf
> Rocoto reported that the build failed for gfs_utils
> .FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
> Rocoto reported that the build failed for gsi_enkf
> Rocoto reported that the build failed for gfs_utils
> .FATAL ERROR: /scratch3/NCEPDEV/global/role.glopara/GFS_CI_CD/URSA/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/global-workflow/sorc/build_compute.sh rocoto failed with state 'FAIL'
> Rocoto reported that the build failed for gsi_enkf
> Rocoto reported that the build failed for gfs_utils
> ```

Disk is full?

zhanglikate commented:
> _C96_gcafs_cycled_ **FAILED** on Hercules (pipeline ID: 6239)
> 
> In directory: `/work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/EXPDIR/C96_gcafs_cycled_4e03e8ad-6239`
> 
> Error Log Files:
> 
> ```
> 
> /work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/COMROOT/C96_gcafs_cycled_4e03e8ad-6239/logs/2021122018/gcafs_atmos_prod_f000-f003.log
> ```
> 
> View Error Logs: [(gcafs_atmos_prod_f000-f003.log)](https://gist.github.com/emcbot/c0e3fa62343b0db4690bc80771366cac)
> 
> _This failure was detected automatically by global-workflow's CI/CD Pipeline_

When do the regression test the config.atmos_products need to be use the the one from gcafs/config.atmos_products folder, can not use the one from gfs/config.atmos_products

zhanglikate commented:
> _C96_gcafs_cycled_ **FAILED** on Hercules (pipeline ID: 6239)
> 
> In directory: `/work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/EXPDIR/C96_gcafs_cycled_4e03e8ad-6239`
> 
> Error Log Files:
> 
> ```
> 
> /work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/COMROOT/C96_gcafs_cycled_4e03e8ad-6239/logs/2021122018/gcafs_atmos_prod_f000-f003.log
> ```
> 
> View Error Logs: [(gcafs_atmos_prod_f000-f003.log)](https://gist.github.com/emcbot/c0e3fa62343b0db4690bc80771366cac)
> 
> _This failure was detected automatically by global-workflow's CI/CD Pipeline_

The master file need to be updated to include the gcafs model variables before doing production.

zhanglikate commented:
> _C96C48mx500_S2SW_cyc_gfs_ **FAILED** on Hercules (pipeline ID: 6239)
> 
> In directory: `/work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/EXPDIR/C96C48mx500_S2SW_cyc_gfs_4e03e8ad-6239`
> 
> Error Log Files:
> 
> ```
> 
> /work2/noaa/global/role-global/GFS_CI_CD/HERCULES/BUILDS/GITLAB/pr_cases_4147_4e03e8ad_6239/RUNTESTS/COMROOT/C96C48mx500_S2SW_cyc_gfs_4e03e8ad-6239/logs/2021122018/gfs_wavepostbndpnt.log
> ```
> 
> View Error Logs: [(gfs_wavepostbndpnt.log)](https://gist.github.com/emcbot/a49f1612d69306f3705a512519645e9b)
> 
> _This failure was detected automatically by global-workflow's CI/CD Pipeline_

@aerorahul Can you please let me know how to turn off "paramlistb" in the PR testing and only run "paramlista" as the current GCAFS configurations need to turn off "paramlistb" and the master file does not include the variables for "paramlistb", which will be updated in future PR.

CoryMartin-NOAA commented:
@zhanglikate that is unreasonable to ask from code managers such as @aerorahul , it is the responsibility of developers to ensure that all necessary changes are included in the pull request. It's not even clear to me exactly would need to be changed to fix this, so someone who isn't familiar with GCAFS probably wouldn't know either. Have you ran the GCAFS test cases on your own manually using `generate_workflows.sh`?

zhanglikate commented:
> @zhanglikate that is unreasonable to ask from code managers such as @aerorahul , it is the responsibility of developers to ensure that all necessary changes are included in the pull request. It's not even clear to me exactly would need to be changed to fix this, so someone who isn't familiar with GCAFS probably wouldn't know either. Have you ran the GCAFS test cases on your own manually using `generate_workflows.sh`?

@CoryMartin-NOAA Thank you for the comments. To clarify, this is not a bug in the code that needs to be fixed. The UPP production workflow currently uses two parameter lists, paramlista and paramlistb. In the develop branch of GCAFS, the GCAFS control file for grib2 variables has not been finalized. So the current master GRIB2 output contains only aerosol species and does not include any meteorological fields. All aerosol species are already included in paramlista, and at the moment we are not using paramlistb. This is why the PR test fails—UPP attempts to process an aerosol-only master file using paramlistb, which currently contains only meteorological variables.

I could temporarily update paramlistb by adding one aerosol field to avoid the crash. However, my understanding is that @lipan-NOAA plans to use paramlistb soon for meteorological or pressure-level outputs and update the GCAFS control file for grib2 variables. Because of this planned change, I checked with the UPP code manager @WenMeng-NOAA before modifying anything. She suggested asking @aerorahul whether the PR test could temporarily use an option/configuration to disable paramlistb processing for GCAFS model, so the test can complete successfully now. Once the meteorological fields are finalized, we would re-enable paramlistb in the production workflow.

Please let me know whether this clarification helps or if you would prefer a different approach for this PR. If @aerorahul confirms that it is not feasible to use an option or configuration to disable paramlistb processing for the PR test, I can update the gcafs.fFFF.paramlist.b.txt file by adding one aerosol field so that the test can pass. Thank you.

aerorahul commented:
I think setting `downset=1` in `dev/parm/config/gcafs/config.atmos_products` would only process the parmlista file.

zhanglikate commented:
> I think setting `downset=1` in `dev/parm/config/gcafs/config.atmos_products` would only process the parmlista file.

@aerorahul Thanks very much for the information. I will update parm/config/gcafs/config.atmos_products to temporally turn off the paramlistb until @lipan-NOAA finalize the paramlistb.

zhanglikate committed:
Turn off paramlistb for gcafs

aerorahul committed:
Merge branch 'develop' into PR-upp

aerorahul reviewed (approved):
test passed on Hercules. approving before merging.

zhanglikate commented:
> test passed on Hercules. approving before merging.

@aerorahul  I really appreciate all your help and patience with this PR.
